src/torcontrol.cpp

```diff
@@ -527,15 +527,6 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
-        // Now that we know Tor is running setup the proxy for onion addresses
-        // if -onion isn't set to something else.
-        if (gArgs.GetArg("-onion", "") == "") {
-            CService resolved(LookupNumeric("127.0.0.1", 9050));
-            proxyType addrOnion = proxyType(resolved, true);
-            SetProxy(NET_ONION, addrOnion);
-            SetReachable(NET_ONION, true);
-        }
-
```

Comment:
Isn't this removal sufficient to fix https://github.com/bitcoin/bitcoin/issues/14722?

Reply:
in some sense yes, i would also feel better to just remove this since something like this was anyway somehow without the users explicit consent.
But i guess the main concern was to leak ip, if there is no proxy, but the implementation we had now for years was sub optimal at best.

If we can find there common ground to have by default no outbound Tor socks proxy, i am fine with this to just remove this. :-). 
On the other hand from privacy point it might be good to only generate inbound onion services, if there is a Tor proxy set, to not leak ip.
And maybe do a hint comment to the users to encourage them to define at least -proxy= and write some hints in the docs.

And then do an extra PR for the -onlynet=onion.

Reply:
@vasild  thx for also looking at this. u might also notice that since -listenonion is the default, i am thinking right now, why not just create the service if we have a proxy in first place?

Bitcoin's net is now big enough and we no longer need to create random onions to grow the net, so if -listenonion is removed as default we would have a bit more privacy and a clear torcontrol api.

Reply:
Correct me if I am wrong, if only the above code snippet is removed (no other changes), then the behavior will be:

`-proxy=` provided | `-onion=` provided | outcome
---|---|---
no | no | outgoing tor connections are not possible
no | yes | outgoing tor connections will be made via the proxy provided in `-onion=`
yes | no | outgoing tor connections will be made via the proxy provided in`-proxy=` (`-onion=` defaults to `-proxy=`)
yes | yes | outgoing tor connections will be made via the proxy provided in `-onion=`

regardless of whether `-torcontrol=` is provided or not. Which to me makes perfect sense. Users that want to avoid clearnet outgoing connections should set `-proxy=` to the tor socks5 proxy.

I think fixing `-onlynet=` is a separate issue and warrants a separate PR.

Reply:
Makes sense and i addressed ur suggestion.

We now will not create the default outbound proxy and fix by that also the issue and dump a reminder in log  that the user might leak his ip when no proxy is defined and the service call succeeds later somehow on purpose or i.e. when he enables Tor and the node runs already with listenonion, so at least better than the old footgun.

And since we now do not override -onlynet in torcontrol even in that regard better than what we had since years.

Reply:
> Correct me if I am wrong, if only the above code snippet is removed (no other changes), then the behavior will be:

Thank you for the work of analyzing the behavior, @vasild and creating the overview table, this is very good.

Please see laanwj's destinction of when using names "Tor" and when using "Onion" here:
https://github.com/bitcoin/bitcoin/pull/13575#issuecomment-402427844

As I understand, the `onion=` option is only about "onion" connections (that is, connections to .onion Onion services (aka deprecated name Hidden service).

While the `proxy=` setting, when set to a Tor node, forwards IPv4 (non .onion) requests to the Tor network, when they go out to the clearnet from a Tor exit node. As I underdstand, this would be named "tor connection".

So, e.g. if `proxy` is unset and `onion` is set, then there will be Onion (and maybe IPv4/IPv6) connections, but no Tor connections.

And if `proxy` is set and `onion` is unset, then there will only be Tor connections connections (both via exit node to clearnet and to Onion services).

Would be an experiment to set both `proxy` and `onion` to different tor servers and check if really IPv4 connections go through the proxy-Tor server, and .onion connections go through the onion-Tor server. But I think this is not really relevant as I can not think of a use case of having two Tor servers to separate tor and onion connections.

So I think the definition would be:
- Tor connection: Connection to IPv4 (and IPv6?) clearnet through Tor network tunnel, and optional to .onion services.
- Onion connection: Connection to .onion service.